engine/server.ts

```diff
@@ -23,8 +23,132 @@ const sessionSwitchers = {}; // Should be a persistent store...
 const switcherStatus = {}; // Should be a persistent store...
 const eventStreams = {};
 
-class ChannelEngine {
-  constructor(assetMgr, options) {
+export interface ChannelEngineOpts {
+  defaultSlateUri?: string;
+  slateRepetitions?: number;
+  slateDuration?: number;
+  redisUrl?: string;
+  memcachedUrl?: string;
+  sharedStoreCacheTTL?: number;
+  heartbeat?: string;
+  channelManager: IChannelManager;
+  streamSwitchManager?: IStreamSwitchManager;
+  cacheTTL?: number;
+  playheadDiffThreshold?: number;
+  maxTickInterval?: number;
+  cloudWatchMetrics?: boolean;
+  useDemuxedAudio?: boolean;
+  alwaysNewSegments?: boolean;
+  diffCompensationRate?: number;
+  staticDirectory?: string;
+  averageSegmentDuration?: number;
+  targetDurationPadding?: boolean;
+  forceTargetDuration?: boolean;
+  adCopyMgrUri?: string; // deprecated
+  adXchangeUri?: string; // deprecated
+}
+
+interface StreamerOpts {
+  defaultAverageSegmentDuration?: number;
+  cacheTTL?: number;
+  defaultPlayheadDiffThreshold?: number;
+  defaultMaxTickInterval?: number;
+  targetDurationPadding?: boolean;
+  forceTargetDuration?: boolean;
+  diffCompensationRate?: number;
+}
+
+export interface VodRequest {
+  sessionId: string;
+  category?: string;
+  playlistId: string;
+}
+
+export interface VodResponseMetadata {
+  id: string;
+  title: string;
+}
+
+export interface VodTimedMetadata {
+  'start-date': string;
+}
+
+export interface VodResponse {
+  uri: string;
+  currentMetadata?: VodResponseMetadata;
+  timedMetadata?: VodTimedMetadata;
+}
+
+export interface IAssetManager {
+  getNextVod: (vodRequest: VodRequest) => Promise<VodResponse>;
+}
+
+export interface ChannelProfile {
+  bw: number;
+  codecs: string;
+  resolution: number[];
+}
+
+export interface Channel {
+  id: string;
+  profile: ChannelProfile[];
+  audioTracks?: AudioTracks[];
+}
+
+export interface AudioTracks {
+  language: string;
+  name: string;
+  default?: boolean;
+}
+
+export interface IChannelManager {
+  getChannels: () => Channel[];
+}
+
+export enum ScheduleStreamType {
+  LIVE = 1,
+  VOD = 2
+};
+
+export interface Schedule {
+  eventId: string;
+  assetId: string;
+  title: string;
+  type: ScheduleStreamType;
+  start_time: number;
+  end_time: number;
+  uri: string;
+  duration: number;
```

Comment:
'duration' field is only required for when the ScheduleStreamType is 'VOD'. Can be omitted if 'LIVE'. 
Not sure if better to make it optional or simply let the users set any value when it is live. 
'duration' is for how long the channel should stay on the scheduled VOD event before switching back to V2L. Although, it could be also be calculated through delta of start and end time fields.

Reply:
If the config applies to two different type I would recommend splitting it into a `Schedule` base interface, and a `ScheduleVod` and `ScheduleLive` that extend the interface. Any place the interface is used can be changed to `ScheduleVod | ScheduleLive`

Reply:
Indeed a good idea, but I don't want to risk breaking any backward compatibility (with existing plugins) so I will just flag duration field as optional now.

Reply:
Since merging this branch is already a breaking change (the `import { ChannelEngine }` change) it could be reasonable to have other breaking changes lumped together, but I understand the concern